sensor/common/registry/lazy_tls_factory.go

```diff
@@ -0,0 +1,95 @@
+package registry
+
+import (
+	pkgErrors "github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/registries"
+	"github.com/stackrox/rox/pkg/registries/docker"
+	dockerFactory "github.com/stackrox/rox/pkg/registries/docker"
+	rhelFactory "github.com/stackrox/rox/pkg/registries/rhel"
+	"github.com/stackrox/rox/pkg/registries/types"
+)
+
+var (
+	lazyEligibleCreators = []types.CreatorWrapper{
+		dockerFactory.CreatorWithoutRepoList,
+		rhelFactory.CreatorWithoutRepoList,
+	}
+)
+
+var _ registries.Factory = (*lazyFactory)(nil)
+
+type lazyFactory struct {
+	creators      map[string]types.Creator
+	tlsCheckCache *tlsCheckCacheImpl
+}
+
+func newLazyFactory(tlsCheckCache *tlsCheckCacheImpl) registries.Factory {
+	factory := &lazyFactory{
+		creators:      make(map[string]types.Creator, len(lazyEligibleCreators)),
+		tlsCheckCache: tlsCheckCache,
+	}
+
+	for _, creatorFunc := range lazyEligibleCreators {
+		typ, creator := creatorFunc()
+		factory.creators[typ] = creator
+	}
+
+	return factory
+}
+
+// CreateRegistry performs aggressive up front validation so that errors can
+// be surfaced, otherwise errors may be lost during lazy initialization.
+func (e *lazyFactory) CreateRegistry(source *storage.ImageIntegration, options ...types.CreatorOption) (types.ImageRegistry, error) {
+	if source == nil {
+		return nil, pkgErrors.New("image integration is nil")
+	}
+
+	creator, exists := e.creators[source.GetType()]
+	if !exists {
+		return nil, pkgErrors.Errorf("registry with type '%s' does not exist", source.GetType())
+	}
+
+	dockerConfig, err := extractDockerConfig(source)
+	if err != nil {
+		// Only integrations with a docker config are eligible for lazy tls checking
+		return nil, pkgErrors.Wrap(err, "extracting docker config")
+	}
+
+	hostname, url := docker.RegistryHostnameURL(dockerConfig.GetEndpoint())
+	if hostname == "" {
+		return nil, pkgErrors.New("empty registry hostname")
+	}
+
+	return &lazyTLSCheckRegistry{
+		source:           source,
+		creator:          creator,
+		dockerConfig:     dockerConfig,
+		url:              url,
+		registryHostname: hostname,
+		dataSource: &storage.DataSource{
+			Id:   source.GetId(),
+			Name: source.GetName(),
+		},
+		tlsCheckCache: e.tlsCheckCache,
+	}, nil
+}
+
+func extractDockerConfig(ii *storage.ImageIntegration) (*storage.DockerConfig, error) {
+	protoCfg := ii.GetIntegrationConfig()
+	if protoCfg == nil {
+		return nil, pkgErrors.New("image integration config is nil")
+	}
+
+	cfg, ok := protoCfg.(*storage.ImageIntegration_Docker)
+	if !ok || cfg == nil {
+		return nil, pkgErrors.New("image integration docker config is nil")
+	}
+
+	dockerConfig := cfg.Docker
+	if dockerConfig == nil {
+		return nil, pkgErrors.New("docker config is nil")
+	}
```

Comment:
nit:

```suggestion
	dockerCfg := ii.GetDocker()
	if dockerCfg == nil {
		return nil, pkgErrors.Errorf("image integration %q does not have a Docker configuration", ii.GetName())
	}

	return dockerCfg, nil
```

Or this can just return `*storage.DockerConfig`. Or perhaps this isn't needed since it may be simplified. Unless you want to have those granular `nil` checks

Reply:
TIL. Excellent.